core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedStreamSourceChannel.java

```diff
@@ -646,6 +646,10 @@ public synchronized void close() {
                     closeListeners[i].handleEvent(this);
                 }
             }
+            // UNDERTOW-1639: Close may be called from an I/O thread while a worker is blocked on awaitReadable.
+            // Once the channel is closed, callers must be awoken. Note that this does not check 'waiters' because
+            // the value is not volatile and could result in a race.
+            lock.notifyAll();
```

Comment:
This could live in `channelForciblyClosed`, but I'm not sure there's much downside in calling it for each invocation. This way should be safer.